server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, singleNodeConfig, singleNodeConfig, 42L), ackListener, initialFaultyNodes);
+
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            if (failNodeWhenCommitting == false) {
+                if (remainingActions.decrementAndGet() == injectFaultAt) {
+                    publication.onFaultyNode(n2);
+                }
+                if (e.getKey().equals(n2) == false || randomBoolean()) {
+                    PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                        publication.publishRequest);
+                    e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
+                }
+            } else {
+                if (e.getKey().equals(n2) == false || publicationDidNotMakeItToNode2 == false) {
+                    PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                        publication.publishRequest);
+                    e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
+                }
+            }
+        });
+
+        publication.pendingCommits.entrySet().stream().collect(shuffle()).forEach(e -> {
+            if (failNodeWhenCommitting == false) {
+                nodeResolver.apply(e.getKey()).coordinationState.handleCommit(publication.applyCommit);
+                e.getValue().onResponse(TransportResponse.Empty.INSTANCE);
+            } else {
+                if (e.getKey().equals(n2)) {
+                    // we must fail node before committing for the node, otherwise failing the node is ignored
+                    publication.onFaultyNode(n2);
+                }
+                if (remainingActions.decrementAndGet() == injectFaultAt) {
+                    publication.onFaultyNode(n2);
+                }
+                if (e.getKey().equals(n2) == false || randomBoolean()) {
+                    nodeResolver.apply(e.getKey()).coordinationState.handleCommit(publication.applyCommit);
+                    e.getValue().onResponse(TransportResponse.Empty.INSTANCE);
+                }
+            }
+        });
+
+        // we need to complete publication by failing the node
+        if (failNodeWhenCommitting && publicationDidNotMakeItToNode2 && remainingActions.get() > injectFaultAt) {
+            publication.onFaultyNode(n2);
+        }
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        publication.onFaultyNode(randomFrom(n1, n3)); // has no influence
+
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(1));
+        assertThat(errors.get(0).v1(), equalTo(n2));
```

Comment:
Can we also assert something about the error, and also assert that the successful acks are as expected.